src/Microsoft.OData.Core/ObjectPool.cs

```diff
@@ -0,0 +1,49 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ObjectPool.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.OData
+{
+    internal class ObjectPool<T>
+    {
+        private readonly List<T> _objectsInUse;
+        private readonly List<T> _objectsAvailable;
+        private readonly Func<T> _objectGenerator;
+
+        public ObjectPool(Func<T> objectGenerator)
+        {
+            _objectGenerator = objectGenerator ?? throw new ArgumentNullException(nameof(objectGenerator));
+            _objectsInUse = new List<T>();
+            _objectsAvailable = new List<T>();

```

Comment:
I think you can delay creating the array until the first it object is requested (or returned). Not sure if that will make a difference in practice though since most responses are objects.